Ray.jl/src/runtime.jl

```diff
@@ -173,23 +173,76 @@ end
                      resources::Dict{String,Float64}=Dict("CPU" => 1.0))
     export_function!(FUNCTION_MANAGER[], f, get_current_job_id())
     fd = ray_jll.function_descriptor(f)
-    arg_oids = map(Ray.put, flatten_args(args, kwargs))
+    task_args = serialize_args(flatten_args(args, kwargs))
 
     serialized_runtime_env_info = if !isnothing(runtime_env)
         _serialize(RuntimeEnvInfo(runtime_env))
     else
         ""
     end
 
-    return GC.@preserve args ray_jll._submit_task(fd,
-                                                  arg_oids,
-                                                  serialized_runtime_env_info,
-                                                  resources)
+    GC.@preserve task_args begin
+        return ray_jll._submit_task(fd,
+                                    transform_task_args(task_args),
+                                    serialized_runtime_env_info,
+                                    resources)
+    end
+end
+
+# Adapted from `prepare_args_internal`:
+# https://github.com/ray-project/ray/blob/ray-2.5.1/python/ray/_raylet.pyx#L673
+function serialize_args(args)
+    ray_config = ray_jll.RayConfigInstance()
+    put_threshold = ray_jll.max_direct_call_object_size(ray_config)
+    rpc_inline_threshold = ray_jll.task_rpc_inlined_bytes_limit(ray_config)
+    record_call_site = ray_jll.record_ref_creation_sites(ray_config)
+
+    rpc_address = ray_jll.GetRpcAddress()
+
+    total_inlined = 0
+    task_args = Any[]
+    for arg in args
+        # Note: The Python `prepare_args_internal` function checks if the `arg` is an
+        # `ObjectRef` and in that case uses the object ID to directly make a
+        # `TaskArgByReference`. However, as the `args` here are flattened the `arg` will
+        # always be a `Pair` (or a list in Python). I suspect this Python code path just
+        # dead code so we'll exclude it from ours.
+
+        serialized_arg = serialize_to_bytes(arg)
+        serialized_arg_size = sizeof(serialized_arg)
+        buffer = ray_jll.LocalMemoryBuffer(serialized_arg, serialized_arg_size, true)
+
+        # Inline arguments which are small and if there is room
+        task_arg = if (serialized_arg_size <= put_threshold &&
+            serialized_arg_size + total_inlined <= rpc_inline_threshold)
+
+            total_inlined += serialized_arg_size
+            ray_jll.TaskArgByValue(ray_jll.RayObject(buffer))
+        else
+            oid = ray_jll.put(buffer)
+            # TODO: Add test for populating `call_site`
+            call_site = record_call_site ? sprint(Base.show_backtrace, backtrace()) : ""
+            ray_jll.TaskArgByReference(oid, rpc_address, call_site)
+        end
+
+        push!(task_args, task_arg)
+    end
+
+    return task_args
+end
+
+function transform_task_args(task_args)
+    task_arg_ptrs = StdVector{CxxPtr{ray_jll.TaskArg}}()
+    for task_arg in task_args
+        push!(task_arg_ptrs, CxxRef(task_arg))
```

Comment:
So this is calling our custom `push!` which calls `ray_jll._push_back`. I believe this is specifically due to me defining:
```cpp
void _push_back(std::vector<TaskArg *> &vector, TaskArg &el) {
    vector.push_back(&el);
}
```
instead of:
```cpp
void _push_back(std::vector<TaskArg *> &vector, TaskArg *el) {
    vector.push_back(el);
}
```
I'll update the code so we can just pass in a `CxxPtr` here